### PR TITLE
MF-583  module-spa directory GP is no longer relative to app data dir

### PR DESCRIPTION
**Pull request by brandones**
https://issues.openmrs.org/browse/MF-583

Thanks to @corneliouzbett

### CODE REVIEW

**Review by kipchubett — APPROVED** [review]

**Comment by brandones** [comment]
Thanks @corneliouzbett !

Since @corneliouzbett made most of this change himself, I'm inclined to wait for another review by @dkayiwa , @nkimaina , @mseaton , or someone else (@ibacher ?)

File: omod/src/main/java/org/openmrs/module/spa/SpaConstants.java [hunk 1]
```diff
@@ -15,14 +15,18 @@
  */
 public final class SpaConstants {
 
+	SpaConstants() {}
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/SpaConstants.java]: Shouldn't this be `private` as it was before?

File: omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java [hunk 2]
```diff
@@ -35,33 +38,67 @@
...
+        response.setContentLength((int) file.length());
+        String mimeType = getServletContext().getMimeType(file.getName());
+        response.setContentType(mimeType);
+
+        InputStream is = new FileInputStream(file);
+        try {
+            OpenmrsUtil.copyFile(is, response.getOutputStream());
+        } finally {
+            OpenmrsUtil.closeStream(is);
+        }
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java]:
```suggestion
        try (InputStream is = new FileInputStream(file)) {
            OpenmrsUtil.copyFile(is, response.getOutputStream());
        }
```

File: omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java [hunk 3]
```diff
@@ -85,31 +122,58 @@ protected Resource getResource(HttpServletRequest request) {
...
+
+        //Resolve default index.html
+        if (extractedFile.endsWith("index.htm") || !extractedFile.contains(".")) {
+            extractedFile = "index.html";
+        }
+
+        String realPath = folder.getPath();
+        realPath += "/" + extractedFile;
+        realPath = realPath.replace("/", File.separator);
+
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java]:
```suggestion
        File file = folder.toPath().resolve(extractedFile).toFile();
```

File: omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java [hunk 4]
```diff
@@ -85,31 +122,58 @@ protected Resource getResource(HttpServletRequest request) {
...
+        File file = new File(realPath);
+        if (!file.exists()) {
+            log.warn("File with path '" + realPath + "' doesn't exists");
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java]:
```suggestion
            log.warn("File with path '{}' doesn't exist", file.toString());
```

File: omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java [hunk 5]
```diff
@@ -49,15 +49,24 @@ public static File getSpaStaticFilesDir () {
...
+    public static String getRemoteAssetsUrl() {
         AdministrationService administrationService = Context.getAdministrationService();
-        String path = administrationService.getGlobalProperty(SpaConstants.FRONTEND_DIRECTORY_PATH, "/frontend");
+        //Defaults to SPA environment at least for now.
+        String path = administrationService.getGlobalProperty(SpaConstants.GP_REMOTE_URL,
+                "https://spa-modules.nyc3.digitaloceanspaces.com/@openmrs/esm-app-shell/latest/");
 
         if (!path.endsWith("/")) {
-            path = path + "/";
+            path = String.format("%s/", path);
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java]: This seems a little silly to do when `path + "/"` is probably more efficient and easier to read.

File: omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java [hunk 6]
```diff
@@ -35,8 +35,8 @@
      */
     public static File getSpaStaticFilesDir () {
         AdministrationService as = Context.getAdministrationService();
-        String folderName = as.getGlobalProperty(SpaModuleUtils.GLOBAL_PROPERTY_SPA_STATIC_FILES_DIR,
-                SpaModuleUtils.DEFAULT_FRONTEND_DIRECTORY);
+        String folderName = as.getGlobalProperty(GP_LOCAL_DIRECTORY,
+                DEFAULT_FRONTEND_DIRECTORY);
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java]:
```suggestion
        String folderName = as.getGlobalProperty(GP_LOCAL_DIRECTORY, DEFAULT_FRONTEND_DIRECTORY);
```

File: omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java [hunk 7]
```diff
@@ -49,15 +49,24 @@ public static File getSpaStaticFilesDir () {
...
+    }
+
     /*
      * This should take care for File system, Class path, and URL path
      */
-    public static String getFrontendDirectoryPath() {
+    public static String getRemoteAssetsUrl() {
         AdministrationService administrationService = Context.getAdministrationService();
-        String path = administrationService.getGlobalProperty(SpaConstants.FRONTEND_DIRECTORY_PATH, "/frontend");
+        //Defaults to SPA environment at least for now.
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java]:
```suggestion
        // FIXME Defaults to SPA environment at least for now.
```

**Review by ibacher — COMMENTED** [review]
Really just some minor points. This looks fine to me.

**Review by ibacher — APPROVED** [review]

**Comment by kipchubett** [comment]
@brandones @ibacher is it worth adding a liquibase changeset to delete unused/previous GP rows?

**Comment by ibacher** [comment]
@corneliouzbett Given that this was never used in production, I don't think we need to worry about deleting the old GPs.

File: omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java (hunk 7 above)
Comment (brandones) [reply omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java]: I actually don't know what this comment means. Do you? Or @corneliouzbett ?

Comment (ibacher) [reply omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java]: I took it to be trying to explain why the default value is `https://spa-modules.nyc3.digitaloceanspaces.com/@openmrs/esm-app-shell/latest/` (is that actually a useful remote URL?)

**Comment by ibacher** [comment]
@brandones Would you be able to rebase these changes on the latest `master`? There was a update to Maven recently that broke things when using repos that didn't have HTTPS links.

**Pull request merged**